### PR TITLE
Windows JNI build fixes

### PR DESCRIPTION
**Pull request by koldat**
Fixing compilation, unsatisfied link exceptions (updated list of files that needs to be linked) and warnings for Windows build.
```C++
//MSVC 2015 does not support dynamic arrays like:
  rocksdb::Slice key_parts[jkey_parts_len];
//I have converted to:
  std::vector<rocksdb::Slice> key_parts;
```
Also reusing `free_key_parts` that does the same as `free_key_value_parts` that was removed.

Java elapsedTime unit test increase of sleep to 2 ms. Otherwise it was failing.

### CODE REVIEW

File: java/rocksjni/transaction.cc [hunk 1]
```diff
@@ -631,27 +614,23 @@ void txn_write_kv_parts_helper(JNIEnv* env,
...
+  auto jparts_to_free = std::vector<std::tuple<jbyteArray, jbyte*, jobject>>();
 
   // convert java key_parts/value_parts byte[][] to Slice(s)
   for (jsize i = 0; i < jkey_parts_len; ++i) {
     const jobject jobj_key_part = env->GetObjectArrayElement(jkey_parts, i);
     if (env->ExceptionCheck()) {
       // exception thrown: ArrayIndexOutOfBoundsException
-      free_key_value_parts(env, jkey_parts_len, jkey_parts_to_free,
-                           jvalue_parts_to_free);
+      free_key_parts(env, jparts_to_free);
```
Comment (adamretter) [inline java/rocksjni/transaction.cc]: Can you rename `free_key_parts` to just `free_parts` or similar, as it now deals with both keys and values.

Comment (adamretter) [reply java/rocksjni/transaction.cc]: Also where is that function? I don't see it in your PR

File: java/rocksjni/transaction.cc [hunk 2]
```diff
@@ -1582,7 +1556,7 @@ jbyte Java_org_rocksdb_Transaction_getState(JNIEnv* /*env*/, jobject /*jobj*/,
   }
 
   assert(false);
-  return 0xFF;
+  return static_cast<char>(-1);
```
Comment (adamretter) [inline java/rocksjni/transaction.cc]: Why the change from `0xFF` to `-1`?

File: java/CMakeLists.txt [hunk 3]
```diff
@@ -42,69 +44,103 @@ set(JNI_NATIVE_SOURCES
...
+        rocksjni/transaction_options.cc
         rocksjni/ttl.cc
         rocksjni/write_batch.cc
+        rocksjni/writebatchhandlerjnicallback.cc
         rocksjni/write_batch_test.cc
         rocksjni/write_batch_with_index.cc
-        rocksjni/writebatchhandlerjnicallback.cc
 )
 
 set(NATIVE_JAVA_CLASSES
```
[the remark this answers is not in the document]
Comment (koldat) [reply java/CMakeLists.txt]: I simply put all Java there. It is nightmare to keep this in sync. I am fixing this very often after merge with Master. Classes without native simply does not generate anything. Why to bother with that? Wildcard would be even better.

File: java/rocksjni/transaction.cc (hunk 1 above)
Comment (koldat) [reply java/rocksjni/transaction.cc]: I kept original function free_key_parts. It was there two times so I used first and deleted second. I can rename it. No problem.

File: java/rocksjni/transaction.cc (hunk 2 above)
Comment (koldat) [reply java/rocksjni/transaction.cc]: Msvc complains with constant overflow warning (signed char overflow) . This is only solution I have found to fix that warning.

File: java/CMakeLists.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamretter) [reply java/CMakeLists.txt]: Unfortunately it is something that needs to be manually kept in sync. I did discuss using Wildcards before with a contributor who knows CMake well, and they said that wildcards are a bad idea here (I forget their exact reasoning, but it made sense). Please can you just list the native Java classes.

File: java/rocksjni/transaction.cc (hunk 1 above)
Comment (adamretter) [reply java/rocksjni/transaction.cc]: Ah okay, thanks yes, please rename

**Comment by koldat** [comment]
I have applied review changes and rebased (squashed) into one. @adamretter please confirm you are ok.

**Comment by adamretter** [comment]
@koldat this seems to introduce a CMake problem on Linux and Mac (shown in Travis CI). Can you take a look:
```
make[3]: *** No rule to make target `../java/src/main/java/org/rocksdb/util/TickerType.java', needed by `java/CMakeFiles/rocksdbjni_classes.dir/java_compiled_rocksdbjni_classes'.  Stop.
make[2]: *** [java/CMakeFiles/rocksdbjni_classes.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[  0%] Built target build_version
make[1]: *** [java/CMakeFiles/rocksdbjni.dir/rule] Error 2
make: *** [rocksdbjni] Error 2
```

**Comment by koldat** [comment]
Great catch @adamretter This class was refactored, but was left in my working directory when I was refreshing the list. As usual removed that line and rebased to latest master.
